### PR TITLE
Vault, adding envelope and cipher, also fixes

### PR DESCRIPTION
**Pull request by bcoca**
Adding new envelope and cipher to allow for hardened defaults. Also make things more configurable in the future to prevent the need of a new format. Just updating defaults, should work going forward.

TODO: vault classes have been updated, now need to adapt callers and add tests for new format/cipher interactions.

##### ISSUE TYPE

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

While this does technically add new features/configs, it is still a bugfix as we need to backport hardened defaults to all supported versions. 

In this case security trumps other considerations, but we will try to keep this as backwards compatible as possible.

### CODE REVIEW

File: lib/ansible/plugins/filter/vault.yml [hunk 1]
```diff
@@ -15,6 +15,13 @@ DOCUMENTATION:
       description: Vault secret, the key that lets you open the vault.
       type: string
       required: true
+    version:
+      description:
+        - Version of vault format to be used
+        - If not set, the configured default will be used
+      type: string
+      choices: ['1.1', '1.2', '1.3']
+      version_added: '2.18'
```
Comment (bcoca) [inline lib/ansible/plugins/filter/vault.yml:24]: TODO: change on backport

File: lib/ansible/config/base.yml [hunk 2]
```diff
@@ -1161,11 +1161,31 @@ DEFAULT_VAULT_IDENTITY:
 VAULT_ENCRYPT_SALT:
   name: Vault salt to use for encryption
   default: ~
-  description: 'The salt to use for the vault encryption. If it is not provided, a random salt will be used.'
+  description:
+    - The salt to use for the vault encryption. If it is not provided, a random salt will be used.
+    - DO NOT USE in production settings. Specifying a salt decreases the security of vault,
+      specially if it is reused across vaults.
+    - This option is only available as you might required determinism in your context, like when using CI.
```
Comment (bcoca) [inline lib/ansible/config/base.yml:1168]: TODO: deprecate in followup PR, but leave as is for backports

File: lib/ansible/config/base.yml [hunk 3]
```diff
@@ -1161,11 +1161,31 @@ DEFAULT_VAULT_IDENTITY:
...
+    - It is recommended to use '1.3' as other options have static settings and over time security value has decreased.
+    - The '1.1' and '1.2' versions are still available for environments that have to keep compatiblity with older versions of Ansible Core (<2.14).
+  default: '1.3'
+  choices: [ '1.1', '1.2', '1.3']
+  env: [{name: ANSIBLE_VAULT_VERSION}]
+  ini:
+  - {key: vault_version, section: defaults}
+  cli:
+  - name: vault_version
+  version_added: '2.18'
```
Comment (bcoca) [inline lib/ansible/config/base.yml:1188]: TODO: update version on backports

File: lib/ansible/cli/arguments/option_helpers.py [hunk 4]
```diff
@@ -118,6 +118,24 @@ def inner(value):
...
+def forbidden_characters(nope):
+    ''' ensure option does not have any characters in a list'''
+
+    if not isinstance(nope, list):
+        raise Exception(f"Expected list but got {type(nope)}")
+
+    def inner(value):
+        for x in nope:
+            if x in value:
+                raise Exception(f"Invalid character '{x}' found in '{value}'")
```
Comment (webknjaz) [inline lib/ansible/cli/arguments/option_helpers.py]: perhaps, use a better error? like `ValueError`?

File: lib/ansible/cli/arguments/option_helpers.py [hunk 5]
```diff
@@ -118,6 +118,24 @@ def inner(value):
     return inner
 
 
+def forbidden_characters(nope):
+    ''' ensure option does not have any characters in a list'''
+
+    if not isinstance(nope, list):
+        raise Exception(f"Expected list but got {type(nope)}")
```
Comment (webknjaz) [inline lib/ansible/cli/arguments/option_helpers.py]:
```suggestion
        raise TypeError(f"Expected list but got {type(nope)}")
```

File: lib/ansible/cli/arguments/option_helpers.py (hunk 4 above)
Comment (bcoca) [reply lib/ansible/cli/arguments/option_helpers.py]: i was using AnsibleOptionsError  but that didn't seem to work for argparse, will do ValueError

File: lib/ansible/parsing/vault/__init__.py [hunk 6]
```diff
@@ -1139,50 +1074,101 @@ def _editor_shell_command(self, filename):
...
+        if custom_salt:
+            if not origin.startswith('cli:'):
+                display.warning('Using configured salt, reusing a salt across vaults is insecure')
+        else:
+            # Future: replace with call to ansible.utils.encrypt.random_salt(32)
+            custom_salt = os.urandom(32)
+        return to_bytes(custom_salt)
+
+    @staticmethod
+    def _is_equal(b_a, b_b):
```
Comment (mattclay) [inline lib/ansible/parsing/vault/__init__.py:1101]: Unless I've overlooked something, this function is unused, and has been for some time (outside of unit tests).

Comment (bcoca) [reply lib/ansible/parsing/vault/__init__.py:1101]: i just moved the existing one, i did not check for usage in current nor any of the backport targets

Comment (bcoca) [reply lib/ansible/parsing/vault/__init__.py:1101]: last used in 2.11 (outside tests)
```
lib/ansible/parsing/vault/__init__.py:        if not cls._is_equal(b_crypted_hmac, to_bytes(hmac_decrypt.hexdigest())):
```

**Comment by bcoca** [comment]
in favor of new approach in #83697

**Pull request closed without merging**